Correct group controls depth

tonyhallett opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [x] Code is up-to-date with the `master` branch
* [x] Your changes are covered by tests (if possible)
* [ ] You've run `yarn change` locally


PR flow tips:
* [x] Try to start with a Draft PR
* [ ] Once you're ready (ideally the pipeline is passing) promote your PR to Ready for Review. This step will auto-assign reviewers for your PR.
-->

## Current Behavior

[Bug]: DetailsList and GroupedList do not calculate group nesting level correctly - misaligned


## New Behavior

aligned

## Related Issue(s)

<!-- Please link the issue being fixed so it gets closed when this is merged. -->

Fixes #23114

Note that the group footer will need to be looked at in a similar manner

fabricteam commented:
## [Perf Analysis (`@fluentui/react`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>TeachingBubble</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount_base.html">74872</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount.html">74813</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>BaseButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/BaseButton-mount_base.html">702</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/BaseButton-mount.html">680</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Breadcrumb</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Breadcrumb-mount_base.html">2170</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Breadcrumb-mount.html">2187</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>Checkbox</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Checkbox-mount_base.html">2169</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Checkbox-mount.html">2001</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>CheckboxBase</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/CheckboxBase-mount_base.html">1818</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/CheckboxBase-mount.html">1713</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>ChoiceGroup</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ChoiceGroup-mount_base.html">3738</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ChoiceGroup-mount.html">3707</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>ComboBox</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ComboBox-mount_base.html">737</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ComboBox-mount.html">746</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>CommandBar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/CommandBar-mount_base.html">8100</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/CommandBar-mount.html">8038</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>ContextualMenu</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ContextualMenu-mount_base.html">9427</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ContextualMenu-mount.html">9246</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>DefaultButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DefaultButton-mount_base.html">875</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DefaultButton-mount.html">857</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>DetailsRow</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRow-mount_base.html">2797</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRow-mount.html">3028</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>DetailsRowFast</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRowFast-mount_base.html">2913</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRowFast-mount.html">2885</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>DetailsRowNoStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRowNoStyles-mount_base.html">2707</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DetailsRowNoStyles-mount.html">2716</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Dialog</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Dialog-mount_base.html">2497</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Dialog-mount.html">2399</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>DocumentCardTitle</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DocumentCardTitle-mount_base.html">133</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/DocumentCardTitle-mount.html">132</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>Dropdown</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Dropdown-mount_base.html">2500</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Dropdown-mount.html">2595</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FocusTrapZone</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/FocusTrapZone-mount_base.html">1409</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/FocusTrapZone-mount.html">1427</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FocusZone</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/FocusZone-mount_base.html">1327</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/FocusZone-mount.html">1367</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>IconButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/IconButton-mount_base.html">1319</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/IconButton-mount.html">1295</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Label</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Label-mount_base.html">267</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Label-mount.html">270</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Layer</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Layer-mount_base.html">3466</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Layer-mount.html">3408</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Link</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Link-mount_base.html">347</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Link-mount.html">354</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MenuButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/MenuButton-mount_base.html">1107</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/MenuButton-mount.html">1110</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MessageBar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/MessageBar-mount_base.html">1638</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/MessageBar-mount.html">1606</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Nav</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Nav-mount_base.html">2540</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Nav-mount.html">2642</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>OverflowSet</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/OverflowSet-mount_base.html">834</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/OverflowSet-mount.html">863</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Panel</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Panel-mount_base.html">1998</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Panel-mount.html">1928</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Persona-mount_base.html">770</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Persona-mount.html">822</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>Pivot</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Pivot-mount_base.html">1076</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Pivot-mount.html">1080</a></td>
            <td>1000</td>
            <td></td>
           </tr>
<tr>
            <td>PrimaryButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/PrimaryButton-mount_base.html">957</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/PrimaryButton-mount.html">981</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Rating</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Rating-mount_base.html">5767</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Rating-mount.html">5954</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SearchBox</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SearchBox-mount_base.html">1012</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SearchBox-mount.html">1003</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Shimmer</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Shimmer-mount_base.html">2273</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Shimmer-mount.html">2267</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Slider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Slider-mount_base.html">1476</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Slider-mount.html">1624</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SpinButton-mount_base.html">3823</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SpinButton-mount.html">3798</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Spinner</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Spinner-mount_base.html">341</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Spinner-mount.html">352</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SplitButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SplitButton-mount_base.html">2329</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SplitButton-mount.html">2360</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Stack</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Stack-mount_base.html">389</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Stack-mount.html">368</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>StackWithIntrinsicChildren</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/StackWithIntrinsicChildren-mount_base.html">1699</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/StackWithIntrinsicChildren-mount.html">1695</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>StackWithTextChildren</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/StackWithTextChildren-mount_base.html">3935</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/StackWithTextChildren-mount.html">3855</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchColorPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SwatchColorPicker-mount_base.html">9041</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/SwatchColorPicker-mount.html">8982</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>TagPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TagPicker-mount_base.html">1970</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TagPicker-mount.html">2101</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>TeachingBubble</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount_base.html">74872</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount.html">74813</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TeachingBubble-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>Text</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Text-mount_base.html">309</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Text-mount.html">312</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>TextField</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TextField-mount_base.html">1062</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/TextField-mount.html">1053</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>ThemeProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-mount_base.html">965</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-mount.html">978</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>ThemeProvider</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-virtual-rerender_base.html">657</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-virtual-rerender.html">579</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>ThemeProvider</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-virtual-rerender-with-unmount_base.html">1536</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/ThemeProvider-virtual-rerender-with-unmount.html">1585</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Toggle</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Toggle-mount_base.html">602</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/Toggle-mount.html">597</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>buttonNative</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/buttonNative-mount_base.html">100</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24814/perf-test/buttonNative-mount.html">121</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9423"/>

spmonahan commented:
@tonyhallett The issue with API extractor is that you need to commit an updated API snapshot. You can build the snapshot with

```
yarn lage @fluentui/react build
```

Then just commit it and push it.

tonyhallett commented:
I did. Commit "API update"

tonyhallett commented:
This was before I pulled master so perhaps the logic has changed

spmonahan commented:
> This was before I pulled master so perhaps the logic has changed

Ah, you might need to regenerate the API snapshot then. Probably there are other API changes that your branch is missing.

tonyhallett commented:
@spmonahan 
the change to the api snapshot was one that I had introduced in my first commit so I assume that api extractor config had changed.
Snapshot also updated to reflect master.

If the intent agrees then the group footers will need to be changed in the same manner.

spmonahan commented:
@tonyhallett Let's go ahead with these alignment changes.

tonyhallett commented:
@spmonahan 
This pr now addresses #24837 

I think that the code is ok but will compare the examples by eye tomorrow.

spmonahan commented:
> @spmonahan This pr now addresses #24837
> 
> I think that the code is ok but will compare the examples by eye tomorrow.

Sounds good! Before you do that you should probably resolve the merge conflict against master since there are changes to DetailsList. I don't think they'll affect your work but best to find out sooner rather than later :)

I'll be available to take a look next week.

tonyhallett commented on packages/react/etc/react.api.md:
```diff
@@ -6136,6 +6140,8 @@ export type IGroupFooterStyleProps = Required<Pick<IGroupFooterProps, 'theme'>>
+    title: IStyle;
 }
 
 // @public (undocumented)
```
I believe that there should be additional typescript changes.

IDetailsItemProps has groupNestingDepth defined as "Nesting depth of a grouping" yet

IDetailsHeaderProps and IDetailsFooterProps both extend IDetailsItemProps and the DetailsList supplies groupNestingDepth as the "total nesting depth" when calling the render functions.  ( DetailsHeader then aligns ) 

IDetailsGroupDividerProps also extends and again DetailsList supplies "total nesting depth" for the custom renderers of the group header and footer. ( The groupNestingDepth is not used by the GroupHeader and GroupFooter )


----
IDetailsRowProps also extends - it is the "level of the group" which is used for aria and as such agrees with the comment. ( The GroupedListSection passes this to the GroupedList onRenderCell prop function.  DetailsList will call onRenderRow with IDetailsRowProps groupNestingDepth set to this value )

tonyhallett commented on packages/react/etc/react.api.md:
```diff
@@ -1812,6 +1812,9 @@ export { getGlobalClassNames }
 export const GetGroupCount: (groups: IGroup[] | undefined) => number;
 
+// @public
+export const getGroupNestingDepth: (groups: IDetailsListProps['groups']) => number;
```
this can now be made private to DetailsList

tonyhallett commented on packages/react/etc/react.api.md:
```diff
@@ -4839,6 +4842,7 @@ export interface IDetailsRowBaseProps extends Pick<IDetailsListProps, 'onRenderI
     rowFieldsAs?: IComponentAs<IDetailsRowFieldsProps>;
     styles?: IStyleFunctionOrObject<IDetailsRowStyleProps, IDetailsRowStyles>;
     theme?: ITheme;
+    totalNestingDepth?: number;
```
this facilitates the alignment of the items and defaults to the groupNestingDepth for when `<DetailsRow>` is used for rendering by a `<GroupedList>`

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -29,12 +29,14 @@ import { DragDropHelper } from '../../DragDrop';
 import { List, ScrollToMode } from '../../List';
 import { withViewport } from '../../utilities/decorators/withViewport';
-import { GetGroupCount } from '../../utilities/groupedList/GroupedListUtility';
+import { GetGroupCount, getGroupNestingDepth } from '../../utilities/groupedList/GroupedListUtility';
```
can be moved into this file

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -197,7 +199,7 @@ const DetailsListInner: React.ComponentType<IDetailsListInnerProps> = (
   const rowId = getId('row');
 
-  const groupNestingDepth = getGroupNestingDepth(groups);
+  const totalNestingDepth = getGroupNestingDepth(groups);
```
name change to match that of the corresponding prop of DetailsRow

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -340,29 +342,31 @@ const DetailsListInner: React.ComponentType<IDetailsListInnerProps> = (
     return onRenderDetailsGroupFooter
       ? (groupFooterProps: IGroupDividerProps, defaultRender?: IRenderFunction<IGroupDividerProps>) => {
+          const spacingRenderFunction: IRenderFunction<IGroupDividerProps> = originalFooterProps => {
+            originalFooterProps!.groupLevel = totalNestingDepth - 1;
```
this adjustment is for the alignment of footer

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -444,13 +448,38 @@ const DetailsListInner: React.ComponentType<IDetailsListInnerProps> = (
+
+          return (
+            <>
+              <GroupSpacer indentWidth={renderTitleProps?.indentWidth} count={spacing} />
```
alignment of the header

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -489,6 +518,7 @@ const DetailsListInner: React.ComponentType<IDetailsListInnerProps> = (
         compact,
         columns: adjustedColumns,
         groupNestingDepth: nestingDepth,
+        totalNestingDepth: totalNestingDepth,
```
alignment of the items to the header columns

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.base.tsx:
```diff
@@ -1090,15 +1121,7 @@ export class DetailsListBase extends React.Component<IDetailsListProps, IDetails
 
   private _getGroupNestingDepth(): number {
     const { groups } = this.props;
-    let level = 0;
```
algorithm replaced as incorrect

tonyhallett commented on packages/react/src/components/DetailsList/DetailsList.test.tsx:
```diff
@@ -1061,4 +1061,56 @@ describe('DetailsList', () => {
     const groupNameId = checkbox.getAttribute('aria-labelledby')?.split(' ')[1];
     expect(container.querySelector(`#${groupNameId} span`)!.textContent).toEqual('Group 0');
   });
+
```
This does not have a group footer rendering.  Will add if happy with GroupFooter.

May also be worth having a group with no child groups....

tonyhallett commented on packages/react/src/components/DetailsList/DetailsRow.base.tsx:
```diff
@@ -214,6 +214,11 @@ export class DetailsRowBase extends React.Component<IDetailsRowBaseProps, IDetai
+
+    let totalNestingDepth = this.props.totalNestingDepth;
+    if (totalNestingDepth === undefined) {
+      totalNestingDepth = groupNestingDepth;
```
fallback for when used inside by onRenderCell of GroupedList.  A separate totalNestingDepth is not required there as no need to align with any headers.

tonyhallett commented on packages/react/src/components/DetailsList/DetailsRow.base.tsx:
```diff
@@ -363,7 +368,7 @@ export class DetailsRowBase extends React.Component<IDetailsRowBaseProps, IDetai
           indentWidth={indentWidth}
           role="gridcell"
-          count={groupNestingDepth! - (this.props.collapseAllVisibility === CollapseAllVisibility.hidden ? 1 : 0)}
+          count={totalNestingDepth! - (this.props.collapseAllVisibility === CollapseAllVisibility.hidden ? 1 : 0)}
```
the alignment

tonyhallett replied:
I did not change the RowFields columnStartIndex.  ( The default DetailsRowFields does not use it )
```
            <RowFields
              rowClassNames={this._rowClassNames}
              rowHeaderId={`${id}-header`}
              columns={[columnMeasureInfo.column]}
              item={item}
              itemIndex={itemIndex}
              columnStartIndex={(showCheckbox ? 1 : 0) + (groupNestingDepth ? 1 : 0) + columns.length} //***********
              onRenderItemColumn={onRenderItemColumn}
              getCellValueKey={getCellValueKey}
``

tonyhallett commented on packages/react/src/components/GroupedList/GroupFooter.base.tsx:
```diff
@@ -12,8 +13,9 @@ export const GroupFooterBase: React.FunctionComponent<IGroupFooterProps> = props
   if (group && footerText) {
     return (
       <div className={classNames.root}>
+        {props.selectionMode !== SelectionMode.none && <GroupSpacer indentWidth={CHECK_CELL_WIDTH} count={1} />}
```
account for the check if necessary

tonyhallett replied:
className for padding for alignment

tonyhallett commented on packages/react/src/components/GroupedList/GroupedList.base.tsx:
```diff
@@ -231,7 +231,6 @@ export class GroupedListBase extends React.Component<IGroupedListProps, IGrouped
     const headerProps: IGroupHeaderProps = { ...groupProps!.headerProps, ...dividerProps };
     const showAllProps: IGroupShowAllProps = { ...groupProps!.showAllProps, ...dividerProps };
     const footerProps: IGroupFooterProps = { ...groupProps!.footerProps, ...dividerProps };
-    const groupNestingDepth = this._getGroupNestingDepth();
```
no need for calculation

tonyhallett commented on packages/react/src/components/GroupedList/GroupedList.base.tsx:
```diff
@@ -252,7 +251,7 @@ export class GroupedListBase extends React.Component<IGroupedListProps, IGrouped
         group={group}
         groupIndex={groupIndex}
-        groupNestingDepth={groupNestingDepth}
+        groupNestingDepth={1}
```
For sub-groups this value comes from `group.level'.  So strange that for the first level of grouping it did amount to the deepest group level ( or what should have been given that the algorithm was incorrect ).

tonyhallett commented on packages/react/src/components/GroupedList/GroupedList.base.tsx:
```diff
@@ -311,19 +310,6 @@ export class GroupedListBase extends React.Component<IGroupedListProps, IGrouped
     return 'group-' + (group && group.key ? group.key : String(index));
   }
 
-  private _getGroupNestingDepth(): number {
```
no longer necessary as 1 is the correct value

tonyhallett commented on packages/react/src/utilities/groupedList/GroupedListUtility.tsx:
```diff
@@ -24,3 +25,28 @@ export const GetGroupCount = (groups: IGroup[] | undefined): number => {
+ * Takes an array of groups and returns the total depth.
+ * @param groups - The array of groups to walk.
+ */
+export const getGroupNestingDepth = (groups: IDetailsListProps['groups']): number => {
```
The corrected algorithm which can now be moved in to DetailsList

tonyhallett commented on packages/react-examples/src/react/DetailsList/DetailsList.Grouped.Example.tsx:
```diff
@@ -118,10 +148,17 @@ export class DetailsListGroupedExample extends React.Component<{}, IDetailsListG
   }
 
+  private _onRenderGroupFooter: IDetailsGroupRenderProps['onRenderFooter'] = (detailsFooterProps, defaultRender) => {
+    detailsFooterProps!.footerText = 'A footer';
```
there a few places where I used non-null assertion operator and should have checked instead.

tonyhallett commented on packages/react/src/components/DetailsList/DetailsRow.types.ts:
```diff
@@ -241,6 +241,11 @@ export interface IDetailsRowBaseProps
+  /**
+   * Total nesting depth.  For spacing
+   */
+  totalNestingDepth?: number;
```
perhaps should mention that it defaults to groupNestingDepth

micahgodbolt commented:
Is this PR still in progress or should we close it?

tonyhallett commented:
It is still open. There needs to be some input from the fluentui team with regards to expected behaviour of headers and footers. The associated issue has not even been tagged as a bug when it obviously is so.
